yw_website/yw_website/apps/website/serializers.py

```diff
@@ -157,25 +282,186 @@ def __create_tags(self, w, validated_data):
         for script in scripts.data:
-            s = Script(name=script.get('name'), version=v, checksum=script.get('checksum'), content=script.get('content'))
+            s = Script(
+                name=script.get("name"),
+                version=v,
+                checksum=script.get("checksum"),
+                content=script.get("content"),
+            )
             s.save()
-    
-    def __create_files(self, r, validated_data):
-        files = FileSerializer(validated_data.get('files'), many=True)
+
+    def _create_files(self, r, validated_data):
+        files = FileSerializer(validated_data.get("files"), many=True)
         for file in files.data:
             f, _ = File.objects.update_or_create(
-                checksum=file.get('checksum'),
-                defaults={                
-                    'size':file.get('size'),
-                    'name':file.get('name'),
-                    'uri':file.get('uri'),
-                    'last_modified':file.get('lastModified')
-                }
+                checksum=file.get("checksum"),
+                defaults={
+                    "size": file.get("size"),
+                    "name": file.get("name"),
+                    "uri": file.get("uri"),
+                    "last_modified": file.get("lastModified"),
+                },
             )
             f.save()
             rf = RunFile(run=r, file=f)
             rf.save()
 
+    def _create_program_blocks(self, r, validated_data):
+        program_blocks = ProgramBlockSerializer(
+            validated_data.get("programBlock"), many=True, context={"run": r}
+        )
+        for program_block in program_blocks.data:
+            if program_block.get('inProgramBlock') != None:
+                try:
+                    in_block = ProgramBlock.objects.get(
+                        programblock_id=program_block.get("inProgramBlock"), run=r
+                    )
+                except ObjectDoesNotExist:
+                    raise serializers.ValidationError("Parent programblock does not exist when trying to create programBlock")
+            else:
+                in_block = None
+            pb = ProgramBlock(
+                programblock_id=program_block.get("programBlockId"),
+                name=program_block.get("name"),
+                qualified_name=program_block.get("qualifiedName"),
+                in_program_block=in_block,
+                run=r,
+            )
+            pb.save()
+
+    def _create_data(self, r, validated_data):
+        datas = DataSerializer(
+            validated_data.get("data"), many=True, context={"run": r}
+        )
+        for data in datas.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=data.get("inProgramBlock"), run=r
+                )
+            except ObjectDoesNotExist:
+                raise serializers.ValidationError("inProgramBlock not found in Data")
```

Comment:
there can be cases where inProgram block might be null for data, (reference the spec)

Reply:
we should not raise exception here, we should set in_block to null

Reply:
Carl i told you this one was good!

Reply:
Fixed